### ISSUE TITLE
useDialog autoFocus behavior fails in Storybook example for Modal unmounting trigger example.

### ISSUE THREAD

**Issue by majornista** [adobe/react-spectrum#3128]
# 🐛 Bug Report 

This bug was identified when testing #2445. 

Focus gets lost to the hidden underlying content when a secondary modal opens from a popover that closes when the trigger is clicked.

1. Open chrome://settings/accessibility and enable "Show a quick highlight on the focused object."
2. Open https://reactspectrum.blob.core.windows.net/reactspectrum/76d1f51110088049f05a0eb15d941ce2368d0cb2/storybook/index.html?path=/story/modal--unmounting-trigger
3. Tab to focus the Open Popover button within the Storybook example.
4. Press Space key to open the Popover
5. Chrome's focus highlight should render around the Popover dialog itself.
6. Tab to focus the Open modal button within the Popover.
7. Press Space key to open the Modal
8. The Popover will close in the background.

## 🤔 Expected Behavior

9. Chrome's focus highlight should render around the modal Dialog that opens to replace the Popover when the Dialog receives focus.

## 😯 Current Behavior

9. Chrome's focus highlight does not render around the modal Dialog that opens to replace the Popover, because focus is lost to the document.body.

## 💁 Possible Solution

Replace https://github.com/adobe/react-spectrum/blob/9186a10898028b90a3cd4d48c3b0f96f77982fca/packages/%40react-aria/dialog/src/useDialog.ts#L44

with: 

```tsx
        if (document.activeElement === ref.current || !ref.current.contains(document.activeElement)) {
```

So that if the `ref.current` still does not contain have focus after the `setTimeout` duration, the `useDialog` hook will try to focus it safely.

## 🔦 Context

I identified this as a problem while working on a fix for #2444 in #2445.

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| react-spectrum   |  @react-aria/overlays@3.8.2
| Browser          | Google Chrome Version 101.0.4951.54 (Official Build) (arm64)
| Operating System | macOS 12.3.1 (21E258)

## 🧢 Your Company/Team

Adobe/Accessibility
